packages/engine/src/lib/services/storage.service.ts

```diff
@@ -51,24 +51,24 @@ export const createStorageService = ({ workerToken }: { workerToken: string }) =
     }
 }
 
-export function createContextStore({ prefix, flowId, workerToken }: { prefix: string, flowId: FlowId, workerToken: string }): Store {
+export function createContextStore({ prefix, flowId, workerToken, runId }: { prefix: string, flowId: FlowId, workerToken: string, runId: FlowRunId }): Store {
```

Comment:
I would make this file generic `<T>` that returns `Store<T>`, and then I can pass one enum with `FLOW` and `PROJECT` into triggers and another with `runs` as well to the action.

Reply:
Could you please give me an example :sweat_smile:. I didn't get the part about making the 'file' generic.

Reply:
```ts
export interface Store<SCOPE> {
  put<T>(key: string, value: T, scope?: SCOPE): Promise<T>;
  get<T>(key: string, scope?: SCOPE): Promise<T | null>;
  delete(key: string, scope?: SCOPE): Promise<void>;
}

export enum StoreScope {
  // Collection were deprecated in favor of project
  PROJECT = 'COLLECTION',
  FLOW = 'FLOW',
  RUN = 'RUN',
}

```

And then you need to change the create context store

```ts
export function createContextStore<SCOPE extends StoreScope>({ prefix, flowId, workerToken, runId, defaultScope }: { prefix: string, flowId: FlowId, workerToken: string, runId: FlowRunId,
defaultScope: SCOPE  }): Store<SCOPE> {
    return {
        async put<T>(key: string, value: T, scope = defaultScope): Promise<T> {
            const modifiedKey = createKey(prefix, scope, flowId, key, runId)
            await createStorageService({ workerToken }).put({
                key: modifiedKey,
                value,
            })
            return value
        },
        async delete(key: string, scope = defaultScope): Promise<void> {
            const modifiedKey = createKey(prefix, scope, flowId, key, runId)
            await createStorageService({ workerToken }).delete({
                key: modifiedKey,
            })
        },
        async get<T>(key: string, scope = defaultScope): Promise<T | null> {
            const modifiedKey = createKey(prefix, scope, flowId, key, runId)
            const storeEntry = await createStorageService({ workerToken }).get(modifiedKey)
            if (storeEntry === null) {
                return null
            }
            return storeEntry.value as T
        },
    }
}

function createKey<SCOPE extends StoreScope>(prefix: string, scope: SCOPE, flowId: FlowId, key: string, runId: FlowRunId): string {
    switch (scope) {
        case StoreScope.RUN:
            return prefix + 'run_' + runId + '/' + key
        case StoreScope.FLOW:
            return prefix + 'flow_' + flowId + '/' + key
        default:
            return prefix + key
    }
}

```

Now you need to pass <Omit<StoreScope, 'RUN'> when creating it for trigger and without for action, so you make it return a type that doesn't accept run in compilation time.

Sorry for bad explanation, If it's not clear Please let me know

Reply:
Quick question: since the generic in `createContextStore` function extends `StoreScope` and "RUN" in `StoreScope` is required,  wouldn't passing `Omit<StoreScope, 'RUN'>` throw an error?

Reply:
You are correct. I would look for something like ActionStoreScope / TriggerStoreScope, or try to differentiate them. I wrote the code quickly above and didn't test it. 

So please take it with a grain of salt. Sorry.

Reply:
@wise-introvert I took look quick seems correct, I will try to give it a detailed review and merge it tmw :D

Reply:
Thanks a ton @abuaboud!